File: src/Runner.test.ts
import { Effect, Exit, Fiber } from "effect";
import { mkdtempSync, rmSync, readFileSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import * as Path from "node:path";
import { describe, expect, it } from "vitest";
import { runKubectl } from "./Runner.js";

const writeExe = (dir: string, name: string, body: string): string => {
  const p = Path.join(dir, name);
  writeFileSync(p, body, { mode: 0o755 });
  return p;
};

describe("runner", () => {
  it("returns stdout on success", async () => {
    const dir = mkdtempSync(Path.join(tmpdir(), "kf-run-"));
    try {
      const fake = writeExe(dir, "kubectl", "#!/bin/sh\necho hello-stdout\n");
      const out = await Effect.runPromise(runKubectl(["get", "pods"], { kubectlPath: fake, timeoutMs: 5000 }));
      expect(out.trim()).toBe("hello-stdout");
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it("reports a missing binary as a spawn error", async () => {
    const err = await Effect.runPromise(
      Effect.match(runKubectl(["get", "pods"], { kubectlPath: "/nonexistent/kubectl", timeoutMs: 5000 }), {
        onFailure: (e) => e,
        onSuccess: () => null,
      }),
    );
    expect(err?._tag).toBe("KubectlSpawnError");
  });

  it("kills a helper that ignores SIGTERM", async () => {
    const dir = mkdtempSync(Path.join(tmpdir(), "kf-run-"));
    try {
      const childFile = Path.join(dir, "child.pid");
      const fake = writeExe(
        dir,
        "kubectl",
        `#!/bin/sh\necho $$ > "${dir}/parent.pid"\n( trap '' TERM; exec sleep 30 ) &\necho $! > "${childFile}"\ntrap '' TERM\nexec sleep 30\n`,
      );
      const start = Date.now();
      const err = await Effect.runPromise(
        Effect.match(runKubectl(["--context", "saved", "get", "pods"], { kubectlPath: fake, timeoutMs: 600 }), {
          onFailure: (e) => e,
          onSuccess: () => null,
        }),
      );
      expect(err).not.toBeNull();
      expect(err?._tag).toBe("KubectlTimeoutError");
      expect(Date.now() - start).toBeLessThan(10000);
      // The helper must be gone: nothing may keep the OIDC lock.
      const pid = Number(readFileSync(childFile, "utf8").trim());
      const deadline = Date.now() + 3000;
      for (;;) {
        let alive = true;
        try {
          process.kill(pid, 0);
        } catch {
          alive = false;
        }
        if (!alive) break;
        if (Date.now() > deadline) throw new Error(`helper child ${pid} still alive after timeout`);
        await new Promise((r) => setTimeout(r, 50));
      }
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it("cleans the group on interruption", async () => {
    const dir = mkdtempSync(Path.join(tmpdir(), "kf-run-"));
    try {
      const pidFile = Path.join(dir, "parent.pid");
      const fake = writeExe(dir, "kubectl", `#!/bin/sh\necho $$ > "${pidFile}"\nexec sleep 30\n`);
      const fiber = Effect.runFork(runKubectl(["get", "pods"], { kubectlPath: fake, timeoutMs: 30000 }));
      await new Promise((r) => setTimeout(r, 500));
      await Effect.runPromise(Fiber.interrupt(fiber));
      const exit = await Effect.runPromise(Fiber.await(fiber));
      expect(Exit.isFailure(exit)).toBe(true);
      const pid = Number(readFileSync(pidFile, "utf8").trim());
      const deadline = Date.now() + 3000;
      for (;;) {
        let alive = true;
        try {
          process.kill(pid, 0);
        } catch {
          alive = false;
        }
        if (!alive) break;
        if (Date.now() > deadline) throw new Error(`kubectl child ${pid} survived interruption`);
        await new Promise((r) => setTimeout(r, 50));
      }
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });
});

File: internal/kubectl/runner_test.go
package kubectl

import (
	"context"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"syscall"
	"testing"
	"time"
)

func writeExe(t *testing.T, dir, name, body string) string {
	t.Helper()
	p := filepath.Join(dir, name)
	if err := os.WriteFile(p, []byte(body), 0o755); err != nil {
		t.Fatal(err)
	}
	return p
}

func TestRunSuccess(t *testing.T) {
	dir := t.TempDir()
	fake := writeExe(t, dir, "kubectl", "#!/bin/sh\necho hello-stdout\n")
	r := Runner{KubectlPath: fake}
	out, err := r.Run(context.Background(), "get", "pods")
	if err != nil {
		t.Fatalf("Run: %v", err)
	}
	if strings.TrimSpace(out) != "hello-stdout" {
		t.Fatalf("stdout = %q", out)
	}
}

func TestRunTimeoutKillsHelperIgnoringTERM(t *testing.T) {
	dir := t.TempDir()
	childFile := filepath.Join(dir, "child.pid")
	// Fake kubectl spawns a helper child that ignores SIGTERM (like a stuck
	// OIDC credential helper holding the cache lock), then waits. Both must
	// die on timeout via SIGKILL escalation.
	fake := writeExe(t, dir, "kubectl", `#!/bin/sh
echo $$ > "`+dir+`/parent.pid"
( trap '' TERM; exec sleep 30 ) &
echo $! > "`+childFile+`"
trap '' TERM
exec sleep 30
`)
	r := Runner{KubectlPath: fake, TermGrace: 200 * time.Millisecond}
	ctx, cancel := context.WithTimeout(context.Background(), 600*time.Millisecond)
	defer cancel()
	start := time.Now()
	_, err := r.Run(ctx, "--context", "saved", "get", "pods")
	elapsed := time.Since(start)
	if err == nil {
		t.Fatal("want timeout error, got nil")
	}
	if !TimeoutError(err) {
		t.Fatalf("want timeout error, got %v", err)
	}
	// Runner must escalate quickly, not wait for the 30s sleeps.
	if elapsed > 10*time.Second {
		t.Fatalf("escalation too slow: %v", elapsed)
	}
	// Child helper must be gone: no descendant may hold the OIDC lock.
	deadline := time.Now().Add(3 * time.Second)
	for {
		data, readErr := os.ReadFile(childFile)
		if readErr != nil {
			t.Fatalf("child pid file missing: %v", readErr)
		}
		pid, convErr := strconv.Atoi(strings.TrimSpace(string(data)))
		if convErr != nil {
			t.Fatalf("bad pid: %v", convErr)
		}
		if err := syscall.Kill(pid, 0); err != nil {
			break // gone (ESRCH) — success
		}
		if time.Now().After(deadline) {
			t.Fatalf("helper child %d still alive after timeout; OIDC lock could be held", pid)
		}
		time.Sleep(50 * time.Millisecond)
	}
}

func TestRunCancelCleansGroup(t *testing.T) {
	dir := t.TempDir()
	fake := writeExe(t, dir, "kubectl", "#!/bin/sh\ntrap '' TERM\nexec sleep 30\n")
	r := Runner{KubectlPath: fake, TermGrace: 100 * time.Millisecond}
	ctx, cancel := context.WithCancel(context.Background())
	go func() {
		time.Sleep(300 * time.Millisecond)
		cancel()
	}()
	_, err := r.Run(ctx, "get", "pods")
	if err == nil {
		t.Fatal("want cancel error")
	}
	if !TimeoutError(err) {
		t.Fatalf("want timeout/cancel error, got %v", err)
	}
}
